src/java.base/share/classes/sun/security/ssl/SSLEngineImpl.java

```diff
@@ -789,17 +789,14 @@ public void closeInbound() throws SSLException {
             // Is it ready to close inbound?
             //
             // No exception if the initial handshake is not started.
-            if (!conContext.isInputCloseNotified &&
-                (conContext.isNegotiated ||
-                    conContext.handshakeContext != null)) {
-
-                throw conContext.fatal(Alert.INTERNAL_ERROR,
+            if (!conContext.isInputCloseNotified && (conContext.isNegotiated
+                    || conContext.handshakeContext != null)) {
+                throw new SSLException(
                         "closing inbound before receiving peer's close_notify");
             }
-
-            conContext.closeInbound();
         } finally {
-            engineLock.unlock();
+                conContext.closeInbound();
+                engineLock.unlock();
```

Comment:
I see that `onContext.closeInbound()` might throw, which would leave the `engineLock` locked and could cause deadlocks down the road. So maybe you should have a nested `try { } finally { }` here to make sure the lock is properly unlocked.

Reply:
+1.

Reply:
Good catch.  Thanks.